src/controllers/scripting/controllerscriptenginebase.cpp

```diff
@@ -0,0 +1,252 @@
+#include "controllers/scripting/controllerscriptenginebase.h"
+
+#include "control/controlobject.h"
+#include "controllers/controller.h"
+#include "controllers/controllerdebug.h"
+#include "controllers/scripting/colormapperjsproxy.h"
+#include "controllers/scripting/legacy/controllerscriptinterface.h"
+#include "errordialoghandler.h"
+#include "mixer/playermanager.h"
+
+ControllerScriptEngineBase::ControllerScriptEngineBase(Controller* controller)
+        : m_bDisplayingExceptionDialog(false),
+          m_pJSEngine(nullptr),
+          m_pController(controller),
+          m_bTesting(false) {
+    // Handle error dialog buttons
+    qRegisterMetaType<QMessageBox::StandardButton>("QMessageBox::StandardButton");
+
+    // Subclasses are responsible for adding paths to watch to m_scriptWatcher
+    // in ther initializeScriptEngine method.
+    connect(&m_scriptWatcher,
+            &QFileSystemWatcher::fileChanged,
+            this,
+            &ControllerScriptEngineBase::reload);
+}
+
+ControllerScriptEngineBase::~ControllerScriptEngineBase() {
+    shutdown();
+}
+
+bool ControllerScriptEngineBase::initialize() {
+    VERIFY_OR_DEBUG_ASSERT(!m_pJSEngine) {
+        return false;
+    }
+
+    // Create the Script Engine
+    m_pJSEngine = new QJSEngine(this);
+
+    QJSValue engineGlobalObject = m_pJSEngine->globalObject();
+
+    QJSValue mapper = m_pJSEngine->newQMetaObject(
+            &ColorMapperJSProxy::staticMetaObject);
+    engineGlobalObject.setProperty("ColorMapper", mapper);
+
+    if (m_pController) {
+        qDebug() << "Controller in script engine is:"
+                 << m_pController->getName();
+
+        ControllerJSProxy* controllerProxy = m_pController->jsProxy();
+
+        // Make the Controller instance available to scripts
+        engineGlobalObject.setProperty(
+                "controller", m_pJSEngine->newQObject(controllerProxy));
+
+        // ...under the legacy name as well
+        engineGlobalObject.setProperty(
+                "midi", m_pJSEngine->newQObject(controllerProxy));
+    }
+
+    // Binary data is passed from the Controller as a QByteArray, which
+    // QJSEngine::toScriptValue converts to an ArrayBuffer in JavaScript.
+    // ArrayBuffer cannot be accessed with the [] operator in JS; it needs
+    // to be converted to a typed array (Uint8Array in this case) first.
+    // This function generates a wrapper function from a JS callback to do
+    // that conversion automatically.
+    m_makeArrayBufferWrapperFunction = m_pJSEngine->evaluate(QStringLiteral(
```

Comment:
Can you move this to controllerscriptenginelegacy? I think it's not used somewhere else.

The new controller engine we should just pass the raw ArrayBuffer. HID mappings might prefer to use a `DataView` instead. We could move the Uint8Array code into the MidiDispatcher code instead.

Reply:
MidiDispatcher will also use this.

Reply:
I what I meant was: how do you get a DataView representation of the buffer for HID mappings?

The midi dispatcher will be implemented in JS, right? So we should do it there instead. The C++ wrapper can be moved into the legacy engine, because we only need it for old mappings.

Reply:
The new API will pass an ArrayBuffer for HID controllers. The legacy HID API requires using a Uint8Array for backwards compatibility.

Reply:
@Holzhaus in future you will have the choice to receive the whole buffer or use a predefines dispatcher that has a js API to allow to register callbacks. In that case only the subscripted part of the data is passed.

Reply:
Yes, that's what I meant, too. I don't get why this cast to Uint8Array has to be part of ControllerScriptEngineBase. It should be part of the ControllerScriptEngineLegacy because we don't need it elsewhere.

For the new engine, the cast can be part of die MidiDispatcher JS implementation (i.e. not in C++). And for HID mappings this is not needed anyway.